### ISSUE TITLE
Add a "choco run" command

### ISSUE THREAD

**Issue by SlugFiller** [chocolatey/choco#2210]
While not all packages are applications, for ones that are, Chocolatey seems content with polluting the Start menu and/or Desktop (personally, when not using Chocolatey, I prefer to use portable, and just run the program using Explorer from a folder of my choice), assuming it's not just a program that auto-runs and sits in the tray. While not often, it's more than once that I installed a program using Chocolatey, then had to struggle to figure out where exactly it is and how to use/run it.

Then I compare it to things like npm or PlayStore (Yes, that's quite a wide spectrum, I'm aware), and the ability to just run a thing you installed without having to look for it, e.g. "choco run chocolateygui".

This would allow packages to specify optional metadata for "how to run". If possible, a package should be able to provide multiple options, for programs that have more than one way to run them/more than one executable. This could be represented by an option such as "choco run --list chocolateygui". This is less important, though, as I couldn't find any practical example for this on my own computer.

**Comment by gep13** [comment]
@SlugFiller I wanted to follow up with the example that you provided to ensure that the functionality that is already offered by Chocolatey is understood.

If you look at the output of the installation of Chocolatey GUI, you will see the following:

![image](https://user-images.githubusercontent.com/1271146/110081809-65ae2a00-7d84-11eb-8e7e-dc97a153e8cb.png)

And it mentions that it has created a shim for the underlying application.  Once installed, you can immediately run the `chocolateygui` command and Chocolatey GUI will open.

A number of packages use a similar approach.  Does this help?

**Comment by SlugFiller** [comment]
Not really. While it's true that chocolateygui specifically can be ran from the command line, it appears to be the only program to do this from the ones I have installed, aside from NodeJS. Obivously, some programs, like the aforementioned NodeJS, naturally run from the command line. But that's different from being executed from Chocolatey. It's just their natural run mode.

ChocolateyGUI having a shim is not the same as a solution. It's non-standard, meaning most packages don't do this, and tools, such as ChocolateyGUI itself, can't rely on it. For example, could you add a "run" button to ChocolateyGUI to run an installed program? Right now, you can't, because there's no machine-readable way for it to detect if the program has a shim.

What I'm talking is a standardized way for a package to specify how the program it installs (if any) should be ran. It should specify this in a machine-readable way. And it should be a standard, to signal to package maintainers that if their package is an application (as opposed to a library or driver), then they should specify this extra metadata.

But most importantly, the ability to rely on "choco install <name>" and "choco run <name>" using the exact same <name> is something you can't guarantee with a shim (since a package can install a shim under any name it pleases).

Also, global-namespacing the shim is ultra bad behavior in general. Like, what if you actually have a program that has the same name as Chocolatey package? Creating a global namespace shim in that case would be really bad behavior.

**Comment by Luuk34** [comment]
I was reading this, and it is really unclear what the difference is between installing, and then starting the program, as two separate actions or as one, if I understand this request correctly.

Can you share an example which clearly shows the benefits of an added feature to choco for the "choco run" command? (I am curious 😉)

**Comment by SlugFiller** [comment]
@Luuk34 What I'm suggesting isn't a one action install-and-run. It would still be two actions. The issue I'm describing has to do with the latter action.

What you're describing as "start the program" is currently a non-standardized action that varies widely between packages, being utterly obscure for many of them. The association between the "install" command, which is a straightforward standard, and the "start the program" process, which could be basically anything, is often so unclear, that you're not even sure if uninstalling a certain package would make a certain program you've been running disappear.

I'm suggesting turning "start the program" from a non-standard process that has nothing to do with Chocolatey, to a command that can be run within Chocolatey, has a single uniform structure that doesn't vary from package to package, and is even machine-readable, allowing third-party tools to be able to automatically know how to run a certain package, a process that would currently require human intervention.

The use case is any time I installed a program, and then had to waste 10-15 minutes looking for it, to figure out what shape and form it actually installed as.

It's about changing from "Every package does its own thing" to "One standard". Does this make it clear?

**Comment by TheCakeIsNaOH** [comment]
> Chocolatey seems content with polluting the Start menu and/or Desktop 

I agree that this is a problem, but I disagree that it is caused by Chocolatey. In the packages that I have used, almost all the desktop and start menu items that are added by an installer that a package runs. So if the installers had been run manually, the desktop would be the same as if they had been installed through choco. So, I think it is an issue that is prevalent to windows in general, but choco could do more to simplify management of these.

I'd suggest following #4, and the linked issues there.

> Also, global-namespacing the shim is ultra bad behavior in general.

I'm not sure what you suggest as the alternative? The point of shimming is to put the executable is on the system path, without adding an extra directory to the path.

> Like, what if you actually have a program that has the same name as Chocolatey package? Creating a global namespace shim in that case would be really bad behavior.

Just run that other program with the full path to it, like you would if neither was installed through choco?

Perhaps you are misunderstanding what shims are:
https://docs.chocolatey.org/en-us/features/shim

> It's non-standard, meaning most packages don't do this

Shims are the standard way to put an executable on the path with choco. There are the occasional packages where shims are not used, but these are definitely the exception.

> The use case is any time I installed a program, and then had to waste 10-15 minutes looking for it, to figure out what shape and form it actually installed as.

During install for many packages, choco will print out a message like `Software installed to 'C:\Program Files\programName'`, in which case you can look in `'C:\Program Files\programName'` for the program. Or if a shim is created during install, it will print out the name of the shim that was created, in which case, run the name of the shim. Or use the Start/Desktop shortcuts that you are complaining about.

For packages that have software that I need to run manually, one of those three will give me what I need in %98 of cases. 

So perhaps paying attention to what choco is printing out during install might fix most of you issue?

> It's about changing from "Every package does its own thing" to "One standard".

But the problem with this is that there is not one type of Chocolatey package.

Yes, some packages do install a program that has a single executable that you may want to run/start from the command line. But there are a bunch of other things a package could do where what to run is not so clear-cut.
- Install a windows updates (which you don't run after install (in most cases))
- Change a registry value or two
- Configure another program
- Virtual packages (packages that only take dependencies on other package(s))
- Install a bundle of programs (sysinternals comes to mind, which program out of the bundle of thirty should be run?)
- Install a program that has multiple different executables for different purposes (java.exe vs javac.exe from a JDK for example)
- Install a program that runs as a background service
- Install a library (either system wide, or for a another program)
- Install a driver
- Install a plugin/addon for another program
- Add a windows component (dotnet3.5)

**Comment by SlugFiller** [comment]
@TheCakeIsNaOH
> I agree that this is a problem, but I disagree that it is caused by Chocolatey.

It's caused by a lack of alternatives, which is caused by Chocolatey. Yes, a package could simply not put anything on the Desktop or start menu (even when a package is based on an installer, the installer can be configured not to create start menu items). But then what? How do you run the program?

It's not that package maintainers fail to avoid creating new icons, it's that they see no alternative to doing that.

> I'm not sure what you suggest as the alternative?

Instead of "myprogram" you run "choco run myprogram". Then, if you already have a program called "myprogram", there's no collision, because the addition of "choco run" namespaces it. "myprogram" runs the original, "choco run myprogram" runs the one installed via chocolatey.

> Perhaps you are misunderstanding what shims are

No, I just object to global namespace pollution. The other alternative is to have a common prefix for shim names, like "choco-myprogram", but that would be functionally the same as "choco run".

> During install for many packages, choco will print out a message like

I often install programs via Chocolatey GUI which does not keep the command line open after installation. And this is before counting batch installations of packages, where such a line would be buried inside the long output of multiple packages. But I'll mentally keep "scan the installation log" as a very user unfriendly option.

> But there are a bunch of other things a package could do where what to run is not so clear-cut.

I prefaced my suggestion with "not all applications are packages". I've also mentioned this case:

>Install a program that has multiple different executables for different purposes (java.exe vs javac.exe from a JDK for example)

I talked about the idea of "having multiple options", but mentioned that I couldn't find any practical examples. Java doesn't really work as an example, as both options require additional command line parameters to do anything. Program bundles would probably work better as an example, assuming there was no way to separate them into different packages.

For drivers, updates and libraries, there's no need to include any option to run. For services, running a configuration tool seems appropriate. Virtual packages may either leave it up to the dependent package, or include all of the options.

Some variance between packages will inevitably exist, as it exists for various things now (e.g. packages that do or don't disable auto updates, or tray icons). But it will be within a smaller, more easily enumerable range.

**Comment by Luuk34** [comment]
> It's about changing from "Every package does its own thing" to "One standard". Does this make it clear?

You comment makes thing (a bit) more clear, but I (personally) still do no see an advantage to this option.

**Comment by realSergiy** [comment]
This could be an option similar to the glorious and heavily used npx
https://www.npmjs.com/package/npx
Choco could have `chx` to just run something.
Example:
```
chx resharper-clt MySolution.sln
```
(downloads and runs code inspection with JetBrains ReSharper Command Line Tools (Portable) and executes without install/put into $PATH)
```
chx firefox
```
(downloads and starts Firefox browser (Portable) without install)

I'd suggest `--no-install` to be a default option.

**Comment by SlugFiller** [comment]
@realSergiy Except `npx` works because `npm` already has a `run` option. If Chocolatey has a `run` option, then yes, it would be possible to create a `chx` as a script that runs `choco install` followed by `choco run`. So rather than "similar", it's more of a "prerequisite".
